demos/bubble_shear.py

```diff
@@ -0,0 +1,483 @@
+metric_adv_error = compute_rel_error(metric_adv_c0, c)
+
+print(
+    f"Metric advection mesh adaptation.\n"
+    f"   Avg. number of vertices: {np.average(mesh_numVertices):.1f}\n"
+    f"   Relative L2 error: {metric_adv_error:.2f}%"
+)
+
+# .. code-block:: console
+#
+#    Metric advection mesh adaptation.
+#       Avg. number of vertices: 2032.7
+#       Relative L2 error: 31.30%
+#
+# .. figure:: bubble_shear-metric_advection_14.jpg
+#    :figwidth: 80%
+#    :align: center
+#
+# The relative L2 error of 31.00% is slightly lower than the one obtained with the
+# classical mesh adaptation algorithm, but note that the average number of vertices is
+# about 15% smaller. Looking into the final adapted mesh and concentration fields in
+# the above figure, we now observe that the mesh is indeed refined in a much wider
+# region - ensuring that the bubble remains well-resolved throughout the subinterval.
+# However, this also means that the available resolution is more widely distributed,
+# leading to a coarser local resolution compared to classical mesh adaptation.
+#
+# Let us now consider the implications and limitations of each approach. Firstly,
+# given a high enough adaptation frequency (i.e. number of adaptations/subintervals),
+# the lagging mesh problem can be mitigated with the classical mesh adaptation
+# algorithm. This may end up being more computationally efficient than the metric
+# advection algorithm, which requires solving the advection equation once more
+# at each timestep, as well as potentially expensive metric computations.
+# However, while increasing the mesh adaptation frequency would alleviate the mesh lag,
+# doing so may introduce large errors due to frequent solution transfers between meshes.
+#
+# This is where the advantage of the metric advection algorithm lies: it predicts where
+# to prescribe fine resolution in the future, and thus avoids the need for frequent
+# solution transfers. We can assure ourselves of that by repeating the above simulations
+# with `num_adaptations = 5`, which yields relative L2 errors of 61.06% and 36.86% for
+# the classical and metric advection algorithms, respectively. Furthermore, the problem
+# considered in this example is relatively well-suited for classical mesh adaptation,
+# as the bubble concentration field reverses and therefore often indeed remains in the
+# finely-resolved region. We can observe that in the below figure, at the subinterval
+# :math:`(1.4 s, 1.6 s]`.
+#
+# .. figure:: bubble_shear-classical_7.jpg
+#    :figwidth: 80%
+#    :align: center
+#
+# In conclusion, the choice of mesh adaptation algorithm depends on the specific problem
+# at hand, as well as the computational resources available. We encourage the reader to
+# experiment with different metric parameters, different adaptation frequencies, and
+# even different velocity fields to further explore the capabilities and limitations of
+# the above-presented mesh adaptation algorithms.
```

Comment:
Restructure this more formally in an exercise section, like in other demos.

Reply:
Done in 414a6a90769add123a14510d67ae217efc7335b0